File: Java Barcode Scanner ME/app/src/main/java/de/dimedis/mobileentry/backend/BackendApi.java
package de.dimedis.mobileentry.backend;

import com.google.gson.JsonObject;

import de.dimedis.mobileentry.backend.response.BatchUploadResponse;
import de.dimedis.mobileentry.backend.response.DownloadLanguagesResponse;
import de.dimedis.mobileentry.backend.response.DownloadMyAvailableBordersResponse;
import de.dimedis.mobileentry.backend.response.DownloadSettingsResponse;
import de.dimedis.mobileentry.backend.response.GetTicketHistoryResponse;
import de.dimedis.mobileentry.backend.response.GetVersionsResponse;
import de.dimedis.mobileentry.backend.response.GrabIdResponse;
import de.dimedis.mobileentry.backend.response.PerformCheckoutResponse;
import de.dimedis.mobileentry.backend.response.PerformEntryResponse;
import de.dimedis.mobileentry.backend.response.RecordEntryResponse;
import de.dimedis.mobileentry.backend.response.ResetDeviceResponse;
import de.dimedis.mobileentry.backend.response.SendHeartbeatResponse;
import de.dimedis.mobileentry.backend.response.ServerConnectResponse;
import de.dimedis.mobileentry.backend.response.StealIdResponse;
import de.dimedis.mobileentry.backend.response.UploadOfflineSessionsResponse;
import de.dimedis.mobileentry.backend.response.UserLoginByBarcodeResponse;
import de.dimedis.mobileentry.backend.response.UserLogoutResponse;
import de.dimedis.mobileentry.backend.response.VoidTicketResponse;
import io.reactivex.rxjava3.core.Observable;
import okhttp3.ResponseBody;
import retrofit2.Response;
import retrofit2.http.Field;
import retrofit2.http.FormUrlEncoded;
import retrofit2.http.POST;
import retrofit2.http.Streaming;

public interface BackendApi {
    String PATH = // "";//
            "Dim_fm::Ctrl::FairMateMobileEntry3";

    String REQUEST_ID = "request_id";
    String LANG = "lang";
    String COMM_KEY = "comm_key";
    String DEVICE_ID = "device_id";
    String DEVICE_SUID = "device_suid";
    String DEVICE_TYPE = "device_type";
    String USER_NAME = "user_name";
    String USER_PASSWORD = "user_password";
    String USER_SUID = "user_suid";
    String USER_SESSION = "user_session";
    String ACTIVE_USER_SESSION = "active_user_session";
    String USER_PREFS = "user_prefs";
    String LOCAL_RECORDS = "local_records";
    String VERSIONS = "versions";
    String SYSTEM_HEALTH = "system_health";
    String FAIR = "fair";
    String BORDER = "border";
    String TICKET_CODE = "ticket_code";
    String LOGIN_BARCODE = "login_barcode";
    String CUSTOMER_TOKEN = "customer_token";
    String DATA = "data";

    @FormUrlEncoded
    @POST(PATH + "/server_connect")
    Observable<ServerConnectResponse> serverConnect(
            @Field(CUSTOMER_TOKEN) String customerToken);

    @FormUrlEncoded
    @POST(PATH + "/grab_id")
    Observable<GrabIdResponse> grabId(
            @Field(DEVICE_ID) String deviceId,
            @Field(LANG) String lang);

    @FormUrlEncoded
    @POST(PATH + "/steal_id")
    Observable<StealIdResponse> stealId(
            @Field(DEVICE_ID) String deviceId,
            @Field(LANG) String lang);

    @FormUrlEncoded
    @POST(PATH + "/user_login_by_barcode")
    Observable<UserLoginByBarcodeResponse> userLoginByBarcodeBackend(
            @Field(LANG) String lang,
            @Field(COMM_KEY) String commKey,
            @Field(DEVICE_SUID) String deviceSuid,
            @Field(LOGIN_BARCODE) String loginBarcode);

    @FormUrlEncoded
    @POST(PATH + "/user_login_by_username_and_password")
    Observable<UserLoginByBarcodeResponse> userLoginByUsernameAndPassword(
            @Field(LANG) String lang,
            @Field(COMM_KEY) String commKey,
            @Field(DEVICE_SUID) String deviceSuid,
            @Field(USER_NAME) String userName,
            @Field(USER_PASSWORD) String userPassword);

    @FormUrlEncoded
    @POST(PATH + "/download_settings")
    Observable<DownloadSettingsResponse> downloadSettings(
            @Field(LANG) String lang,
            @Field(COMM_KEY) String commKey,
            @Field(DEVICE_SUID) String deviceSuid);

    @FormUrlEncoded
    @POST(PATH + "/download_my_available_borders")
    Observable<DownloadMyAvailableBordersResponse> downloadMyAvailableBorders(
            @Field(LANG) String lang,
            @Field(COMM_KEY) String commKey,
            @Field(DEVICE_SUID) String deviceSuid);

    //Response
    @FormUrlEncoded
    @POST(PATH + "/download_offline_config")
    Observable<JsonObject> downloadOfflineConfig(
            @Field(LANG) String lang,
            @Field(COMM_KEY) String commKey,
            @Field(DEVICE_SUID) String deviceSuid);

    @FormUrlEncoded
    @POST(PATH + "/get_versions")
    Observable<GetVersionsResponse> getVersionResponse(
            @Field(LANG) String lang,
            @Field(COMM_KEY) String commKey,
            @Field(DEVICE_SUID) String deviceSuid,
            @Field(DEVICE_TYPE) String deviceType);

    @FormUrlEncoded
    @POST(PATH + "/reset_device")
    Observable<ResetDeviceResponse> resetDevice(
            @Field(LANG) String lang,
            @Field(COMM_KEY) String commKey,
            @Field(DEVICE_SUID) String deviceSuid);

    @FormUrlEncoded
    @POST(PATH + "/send_heartbeat")
    Observable<SendHeartbeatResponse> sendHeartbeat(
            @Field(LANG) String lang,
            @Field(COMM_KEY) String commKey,
            @Field(DEVICE_SUID) String deviceSuid,
            @Field(SYSTEM_HEALTH) String systemHealth,
            @Field(VERSIONS) String versions,
            @Field(USER_NAME) String userName,
            @Field(USER_SUID) String userSuid,
            @Field(LOCAL_RECORDS) String localRecords);

    @FormUrlEncoded
    @POST(PATH + "/get_ticket_history")
    Observable<GetTicketHistoryResponse> getTicketHistory(
            @Field(LANG) String lang,
            @Field(COMM_KEY) String commKey,
            @Field(DEVICE_SUID) String deviceSuid,
            @Field(USER_SESSION) String userSession,
            @Field(USER_SUID) String userSuid,
            @Field(USER_NAME) String userName,
            @Field(FAIR) String fair,
            @Field(BORDER) String border,
            @Field(TICKET_CODE) String ticketCode);

    @FormUrlEncoded
    @POST(PATH + "/perform_entry")
    Observable<PerformEntryResponse> performEntry(
            @Field(REQUEST_ID) String requestId,
            @Field(LANG) String lang,
            @Field(COMM_KEY) String commKey,
            @Field(DEVICE_SUID) String deviceSuid,
            @Field(USER_SESSION) String userSession,
            @Field(USER_SUID) String userSuid,
            @Field(USER_NAME) String userName,
            @Field(FAIR) String fair,
            @Field(BORDER) String border,
            @Field(TICKET_CODE) String ticketCode);

    @FormUrlEncoded
    @POST(PATH + "/perform_checkout")
    Observable<PerformCheckoutResponse> performCheckout(
            @Field(REQUEST_ID) String requestId,
            @Field(LANG) String lang,
            @Field(COMM_KEY) String commKey,
            @Field(DEVICE_SUID) String deviceSuid,
            @Field(USER_SESSION) String userSession,
            @Field(USER_SUID) String userSuid,
            @Field(USER_NAME) String userName,
            @Field(FAIR) String fair,
            @Field(BORDER) String border,
            @Field(TICKET_CODE) String ticketCode);

    @FormUrlEncoded
    @POST(PATH + "/record_entry")
    Observable<RecordEntryResponse> recordEntry(
            @Field(REQUEST_ID) String requestId,
            @Field(LANG) String lang,
            @Field(COMM_KEY) String commKey,
            @Field(DEVICE_SUID) String deviceSuid,
            @Field(USER_SESSION) String userSession,
            @Field(USER_SUID) String userSuid,
            @Field(USER_NAME) String userName,
            @Field(FAIR) String fair,
            @Field(BORDER) String border,
            @Field(TICKET_CODE) String ticketCode);

    @FormUrlEncoded
    @POST(PATH + "/user_logout")
    Observable<UserLogoutResponse> userLogout(
            @Field(LANG) String lang,
            @Field(COMM_KEY) String commKey,
            @Field(DEVICE_SUID) String deviceSuid,
            @Field(USER_SESSION) String userSession,
            @Field(USER_SUID) String userSuid,
            @Field(USER_NAME) String userName,
            @Field(FAIR) String fair,
            @Field(BORDER) String border,
            @Field(USER_PREFS) String userPrefs);

    //batch_upload
    @FormUrlEncoded
    @POST(PATH + "/batch_upload")
    Observable<BatchUploadResponse> batchUpload2(
            @Field(LANG) String lang,
            @Field(COMM_KEY) String commKey,
            @Field(DEVICE_SUID) String deviceSuid,
            @Field(USER_SESSION) String userSession,
            @Field(USER_SUID) String userSuid,
            @Field(USER_NAME) String userName,
            @Field(FAIR) String fair,
            @Field(BORDER) String border,
            @Field(DATA) String ticketCode);

    //upload_offline_sessions
    @FormUrlEncoded
    @POST(PATH + "/upload_offline_sessions")
    Observable<UploadOfflineSessionsResponse> uploadOfflineSessions(
            @Field(LANG) String lang,
            @Field(COMM_KEY) String commKey,
            @Field(DEVICE_SUID) String deviceSuid,
            @Field(ACTIVE_USER_SESSION) String activeUserSession,
            @Field(DATA) String data);

    //batch_upload
    @FormUrlEncoded
    @POST(PATH + "/void_ticket")
    Observable<VoidTicketResponse> voidTicket(
            @Field(LANG) String lang,
            @Field(COMM_KEY) String commKey,
            @Field(DEVICE_SUID) String deviceSuid,
            @Field(USER_SESSION) String userSession,
            @Field(USER_SUID) String userSuid,
            @Field(USER_NAME) String userName,
            @Field(FAIR) String fair,
            @Field(BORDER) String border,
            @Field(TICKET_CODE) String ticketCode);

    //download_library
    @FormUrlEncoded
    @POST(PATH + "/download_library")
    Observable<Response> downloadLibrary(
            @Field(LANG) String lang,
            @Field(COMM_KEY) String commKey,
            @Field(DEVICE_SUID) String deviceSuid);

    //download_library
    @Streaming
    @FormUrlEncoded
    @POST(PATH + "/download_app")
    Observable<Response<ResponseBody>> downloadApp(
            @Field(LANG) String lang,
            @Field(COMM_KEY) String commKey,
            @Field(DEVICE_SUID) String deviceSuid,
            @Field(DEVICE_TYPE) String deviceType);

    @FormUrlEncoded
    @POST(PATH + "/download_languages")
    Observable<DownloadLanguagesResponse> downloadLanguages(
            @Field(LANG) String lang,
            @Field(COMM_KEY) String commKey,
            @Field(DEVICE_SUID) String deviceSuid);
}

File: Kotlin Migration/NTMobileEntry/app/src/main/java/de/dimedis/mobileentry/backend/BackendApi.kt
package de.dimedis.mobileentry.backend

import com.google.gson.JsonObject
import com.google.gson.annotations.SerializedName
import de.dimedis.mobileentry.backend.response.*
import io.reactivex.rxjava3.core.Observable
import okhttp3.ResponseBody
import retrofit2.Response
import retrofit2.http.Field
import retrofit2.http.FormUrlEncoded
import retrofit2.http.POST
import retrofit2.http.Streaming

interface BackendApi {

    @FormUrlEncoded
    @POST(BackendApiConst.PATH + "/server_connect")
    fun serverConnect(
        @Field(BackendApiConst.CUSTOMER_TOKEN) customerToken: String?
    ): Observable<ServerConnectResponse>

    @FormUrlEncoded
    @POST(BackendApiConst.PATH + "/grab_id")
    fun grabId(
        @Field(BackendApiConst.DEVICE_ID) deviceId: String?,
        @Field(BackendApiConst.LANG) lang: String?
    ): Observable<GrabIdResponse>

    @FormUrlEncoded
    @POST(BackendApiConst.PATH + "/steal_id")
    fun stealId(
        @Field(BackendApiConst.DEVICE_ID) deviceId: String?,
        @Field(BackendApiConst.LANG) lang: String?
    ): Observable<StealIdResponse>

    @FormUrlEncoded
    @POST(BackendApiConst.PATH + "/user_login_by_barcode")
    fun userLoginByBarcodeBackend(
        @Field(BackendApiConst.LANG) lang: String?,
        @Field(BackendApiConst.COMM_KEY) commKey: String?,
        @Field(BackendApiConst.DEVICE_SUID) deviceSuid: String?,
        @Field(BackendApiConst.LOGIN_BARCODE) loginBarcode: String?
    ): Observable<UserLoginByBarcodeResponse>

    @FormUrlEncoded
    @POST(BackendApiConst.PATH + "/user_login_by_username_and_password")
    fun userLoginByUsernameAndPassword(
        @Field(BackendApiConst.LANG) lang: String?,
        @Field(BackendApiConst.COMM_KEY) commKey: String?,
        @Field(BackendApiConst.DEVICE_SUID) deviceSuid: String?,
        @Field(BackendApiConst.USER_NAME) userName: String?,
        @Field(BackendApiConst.USER_PASSWORD) userPassword: String?
    ): Observable<UserLoginByBarcodeResponse>

    @FormUrlEncoded
    @POST(BackendApiConst.PATH + "/download_settings")
    fun downloadSettings(
    @Field(BackendApiConst.LANG) lang: String?,
    @Field(BackendApiConst.COMM_KEY) commKey: String?,
    @Field(BackendApiConst.DEVICE_SUID) deviceSuid: String?
    ): Observable<DownloadSettingsResponse>

    @FormUrlEncoded
    @POST(BackendApiConst.PATH + "/download_my_available_borders")
    fun downloadMyAvailableBorders(
        @Field(BackendApiConst.LANG) lang: String?,
        @Field(BackendApiConst.COMM_KEY) commKey: String?,
        @Field(BackendApiConst.DEVICE_SUID) deviceSuid: String?
    ): Observable<DownloadMyAvailableBordersResponse>

    //Response
    @FormUrlEncoded
    @POST(BackendApiConst.PATH + "/download_offline_config")
    fun downloadOfflineConfig(
        @Field(BackendApiConst.LANG) lang: String?,
        @Field(BackendApiConst.COMM_KEY) commKey: String?,
        @Field(BackendApiConst.DEVICE_SUID) deviceSuid: String?
    ): Observable<JsonObject>

    @FormUrlEncoded
    @POST(BackendApiConst.PATH + "/get_versions")
    fun getVersionResponse(
        @Field(BackendApiConst.LANG) lang: String?,
        @Field(BackendApiConst.COMM_KEY) commKey: String?,
        @Field(BackendApiConst.DEVICE_SUID) deviceSuid: String?,
        @Field(BackendApiConst.DEVICE_TYPE) deviceType: String?
    ): Observable<GetVersionsResponse>

    @FormUrlEncoded
    @POST(BackendApiConst.PATH + "/reset_device")
    fun resetDevice(
        @Field(BackendApiConst.LANG) lang: String?,
        @Field(BackendApiConst.COMM_KEY) commKey: String?,
        @Field(BackendApiConst.DEVICE_SUID) deviceSuid: String?
    ): Observable<ResetDeviceResponse>

    @FormUrlEncoded
    @POST(BackendApiConst.PATH + "/send_heartbeat")
    fun sendHeartbeat(
        @Field(BackendApiConst.LANG) lang: String?,
        @Field(BackendApiConst.COMM_KEY) commKey: String?,
        @Field(BackendApiConst.DEVICE_SUID) deviceSuid: String?,
        @Field(BackendApiConst.SYSTEM_HEALTH) systemHealth: String?,
        @Field(BackendApiConst.VERSIONS) versions: String?,
        @Field(BackendApiConst.USER_NAME) userName: String?,
        @Field(BackendApiConst.USER_SUID) userSuid: String?,
        @Field(BackendApiConst.LOCAL_RECORDS) localRecords: String?
    ): Observable<SendHeartbeatResponse>

    @FormUrlEncoded
    @POST(BackendApiConst.PATH + "/get_ticket_history")
    fun getTicketHistory(
        @Field(BackendApiConst.LANG) lang: String?,
        @Field(BackendApiConst.COMM_KEY) commKey: String?,
        @Field(BackendApiConst.DEVICE_SUID) deviceSuid: String?,
        @Field(BackendApiConst.USER_SESSION) userSession: String?,
        @Field(BackendApiConst.USER_SUID) userSuid: String?,
        @Field(BackendApiConst.USER_NAME) userName: String?,
        @Field(BackendApiConst.FAIR) fair: String?,
        @Field(BackendApiConst.BORDER) border: String?,
        @Field(BackendApiConst.TICKET_CODE) ticketCode: String?
    ): Observable<GetTicketHistoryResponse>

    @FormUrlEncoded
    @POST(BackendApiConst.PATH + "/perform_entry")
    fun performEntry(
        @Field(BackendApiConst.REQUEST_ID) requestId: String?,
        @Field(BackendApiConst.LANG) lang: String?,
        @Field(BackendApiConst.COMM_KEY) commKey: String?,
        @Field(BackendApiConst.DEVICE_SUID) deviceSuid: String?,
        @Field(BackendApiConst.USER_SESSION) userSession: String?,
        @Field(BackendApiConst.USER_SUID) userSuid: String?,
        @Field(BackendApiConst.USER_NAME) userName: String?,
        @Field(BackendApiConst.FAIR) fair: String?,
        @Field(BackendApiConst.BORDER) border: String?,
        @Field(BackendApiConst.TICKET_CODE) ticketCode: String?
    ): Observable<PerformEntryResponse>

    @FormUrlEncoded
    @POST(BackendApiConst.PATH + "/perform_checkout")
    fun performCheckout(
        @Field(BackendApiConst.REQUEST_ID) requestId: String?,
        @Field(BackendApiConst.LANG) lang: String?,
        @Field(BackendApiConst.COMM_KEY) commKey: String?,
        @Field(BackendApiConst.DEVICE_SUID) deviceSuid: String?,
        @Field(BackendApiConst.USER_SESSION) userSession: String?,
        @Field(BackendApiConst.USER_SUID) userSuid: String?,
        @Field(BackendApiConst.USER_NAME) userName: String?,
        @Field(BackendApiConst.FAIR) fair: String?,
        @Field(BackendApiConst.BORDER) border: String?,
        @Field(BackendApiConst.TICKET_CODE) ticketCode: String?
    ): Observable<PerformCheckoutResponse>

    @FormUrlEncoded
    @POST(BackendApiConst.PATH + "/record_entry")
    fun recordEntry(
        @Field(BackendApiConst.REQUEST_ID) requestId: String?,
        @Field(BackendApiConst.LANG) lang: String?,
        @Field(BackendApiConst.COMM_KEY) commKey: String?,
        @Field(BackendApiConst.DEVICE_SUID) deviceSuid: String?,
        @Field(BackendApiConst.USER_SESSION) userSession: String?,
        @Field(BackendApiConst.USER_SUID) userSuid: String?,
        @Field(BackendApiConst.USER_NAME) userName: String?,
        @Field(BackendApiConst.FAIR) fair: String?,
        @Field(BackendApiConst.BORDER) border: String?,
        @Field(BackendApiConst.TICKET_CODE) ticketCode: String?
    ): Observable<RecordEntryResponse>

//    @FormUrlEncoded
    @POST(BackendApiConst.PATH + "/user_logout")
    fun userLogout(
        @Field(BackendApiConst.LANG) lang: String?,
        @Field(BackendApiConst.COMM_KEY) commKey: String?,
        @Field(BackendApiConst.DEVICE_SUID) deviceSuid: String?,
        @Field(BackendApiConst.USER_SESSION) userSession: String?,
        @Field(BackendApiConst.USER_SUID) userSuid: String?,
        @Field(BackendApiConst.USER_NAME) userName: String?,
        @Field(BackendApiConst.FAIR) fair: String?,
        @Field(BackendApiConst.BORDER) border: String?,
        @Field(BackendApiConst.USER_PREFS) userPrefs: String?
    ): Observable<UserLogoutResponse>

//    //batch_upload
    @FormUrlEncoded
    @POST(BackendApiConst.PATH + "/batch_upload")
    fun batchUpload2(
    @Field(BackendApiConst.LANG) lang: String?,
    @Field(BackendApiConst.COMM_KEY) commKey: String?,
    @Field(BackendApiConst.DEVICE_SUID) deviceSuid: String?,
    @Field(BackendApiConst.USER_SESSION) userSession: String?,
    @Field(BackendApiConst.USER_SUID) userSuid: String?,
    @Field(BackendApiConst.USER_NAME) userName: String?,
    @Field(BackendApiConst.FAIR) fair: String?,
    @Field(BackendApiConst.BORDER) border: String?,
    @Field(BackendApiConst.DATA) ticketCode: String?
    ): Observable<BatchUploadResponse>

    //upload_offline_sessions
    @FormUrlEncoded
    @POST(BackendApiConst.PATH + "/upload_offline_sessions")
    fun uploadOfflineSessions(
        @Field(BackendApiConst.LANG) lang: String?,
        @Field(BackendApiConst.COMM_KEY) commKey: String?,
        @Field(BackendApiConst.DEVICE_SUID) deviceSuid: String?,
        @Field(BackendApiConst.ACTIVE_USER_SESSION) activeUserSession: String?,
        @Field(BackendApiConst.DATA) data: String?
    ): Observable<UploadOfflineSessionsResponse>

//    //batch_upload
    @FormUrlEncoded
    @POST(BackendApiConst.PATH + "/void_ticket")
    fun voidTicket(
        @Field(BackendApiConst.LANG) lang: String?,
        @Field(BackendApiConst.COMM_KEY) commKey: String?,
        @Field(BackendApiConst.DEVICE_SUID) deviceSuid: String?,
        @Field(BackendApiConst.USER_SESSION) userSession: String?,
        @Field(BackendApiConst.USER_SUID) userSuid: String?,
        @Field(BackendApiConst.USER_NAME) userName: String?,
        @Field(BackendApiConst.FAIR) fair: String?,
        @Field(BackendApiConst.BORDER) border: String?,
        @Field(BackendApiConst.TICKET_CODE) ticketCode: String?
    ): Observable<VoidTicketResponse>

    //download_library
    @FormUrlEncoded
    @POST(BackendApiConst.PATH + "/download_library")
    fun downloadLibrary(
        @Field(BackendApiConst.LANG) lang: String?,
        @Field(BackendApiConst.COMM_KEY) commKey: String?,
        @Field(BackendApiConst.DEVICE_SUID) deviceSuid: String?
    ): Observable<Response<*>>

    //download_library
    @Streaming
    @FormUrlEncoded
    @POST(BackendApiConst.PATH + "/download_app")
    fun downloadApp(
        @Field(BackendApiConst.LANG) lang: String?,
        @Field(BackendApiConst.COMM_KEY) commKey: String?,
        @Field(BackendApiConst.DEVICE_SUID) deviceSuid: String?,
        @Field(BackendApiConst.DEVICE_TYPE) deviceType: String?
    ): Observable<Response<ResponseBody>>

    @FormUrlEncoded
    @POST(BackendApiConst.PATH + "/download_languages")
    fun downloadLanguages(
        @Field(BackendApiConst.LANG) lang: String?,
        @Field(BackendApiConst.COMM_KEY) commKey: String?,
        @Field(BackendApiConst.DEVICE_SUID) deviceSuid: String?
    ): Observable<DownloadLanguagesResponse>
}

data class LoginPostData(
    @SerializedName("UserId") var userID: String,
    @SerializedName("Password") var userPassword: String
)